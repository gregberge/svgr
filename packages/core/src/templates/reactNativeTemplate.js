import { getProps } from './util'

const getComponents = components => {
  if (!components.size) return ''
  const componentsStr = [...components]
    .filter(component => component !== 'Svg')
    .join(', ')
  return `, { ${componentsStr} }`
}

const getWarning = components => {
  if (!components.size) return ''
  const componentList = [...components].join(', ')
  return `
// SVGR has dropped some elements not supported by react-native-svg: ${componentList}
`
}

const reactNativeTemplate = (code, config, state) => {
  const {
    reactNativeSvgReplacedComponents = new Set(),
    unsupportedComponents = new Set(),
  } = state

  const props = getProps(config)
  const components = getComponents(reactNativeSvgReplacedComponents)
  const warnLog = getWarning(unsupportedComponents)

  let result = `import React from 'react'\n`
  result += `import Svg${components} from 'react-native-svg';\n`
  result += warnLog
  result += `const ${state.componentName} = ${props} => ${code}\n\n`
  result += `export default ${state.componentName}`

  return result
}

export default reactNativeTemplate
