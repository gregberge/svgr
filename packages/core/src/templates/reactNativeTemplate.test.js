import reactNativeTemplate from './reactNativeTemplate'

describe('reactNativeTemplate', () => {
  it('should wrap code into a component', () => {
    expect(
      reactNativeTemplate('<Svg />', {}, { componentName: 'Test' }),
    ).toMatchSnapshot()
    expect(
      reactNativeTemplate(
        '<Svg />',
        { titleProp: true },
        { componentName: 'Test' },
      ),
    ).toMatchSnapshot()
    expect(
      reactNativeTemplate('<Svg />', { ref: true }, { componentName: 'Test' }),
    ).toMatchSnapshot()
    expect(
      reactNativeTemplate(
        '<Svg />',
        { expandProps: true },
        { componentName: 'Test' },
      ),
    ).toMatchSnapshot()
    expect(
      reactNativeTemplate(
        '<Svg />',
        { titleProp: true, ref: true },
        { componentName: 'Test' },
      ),
    ).toMatchSnapshot()
    expect(
      reactNativeTemplate(
        '<Svg />',
        { titleProp: true, expandProps: true },
        { componentName: 'Test' },
      ),
    ).toMatchSnapshot()
    expect(
      reactNativeTemplate(
        '<Svg />',
        { ref: true, expandProps: true },
        { componentName: 'Test' },
      ),
    ).toMatchSnapshot()
  })

  it('should import used components', () => {
    expect(
      reactNativeTemplate(
        '<Svg />',
        {},
        {
          componentName: 'Test',
          reactNativeSvgReplacedComponents: new Set(['Group']),
        },
      ),
    ).toMatchSnapshot()
  })

  it('should log unsupported components', () => {
    expect(
      reactNativeTemplate(
        '<Svg />',
        {},
        {
          componentName: 'Test',
          unsupportedComponents: new Set(['Group']),
        },
      ),
    ).toMatchSnapshot()
  })
})
