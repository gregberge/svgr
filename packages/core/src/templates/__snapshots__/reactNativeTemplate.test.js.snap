// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reactNativeTemplate should import used components 1`] = `
"import React from 'react'
import Svg, { Group } from 'react-native-svg';
const Test = () => <Svg />

export default Test"
`;

exports[`reactNativeTemplate should log unsupported components 1`] = `
"import React from 'react'
import Svg from 'react-native-svg';

// SVGR has dropped some elements not supported by react-native-svg: Group
const Test = () => <Svg />

export default Test"
`;

exports[`reactNativeTemplate should wrap code into a component 1`] = `
"import React from 'react'
import Svg from 'react-native-svg';
const Test = () => <Svg />

export default Test"
`;

exports[`reactNativeTemplate should wrap code into a component 2`] = `
"import React from 'react'
import Svg from 'react-native-svg';
const Test = ({ title }) => <Svg />

export default Test"
`;

exports[`reactNativeTemplate should wrap code into a component 3`] = `
"import React from 'react'
import Svg from 'react-native-svg';
const Test = ({ svgRef }) => <Svg />

export default Test"
`;

exports[`reactNativeTemplate should wrap code into a component 4`] = `
"import React from 'react'
import Svg from 'react-native-svg';
const Test = props => <Svg />

export default Test"
`;

exports[`reactNativeTemplate should wrap code into a component 5`] = `
"import React from 'react'
import Svg from 'react-native-svg';
const Test = ({ svgRef, title }) => <Svg />

export default Test"
`;

exports[`reactNativeTemplate should wrap code into a component 6`] = `
"import React from 'react'
import Svg from 'react-native-svg';
const Test = ({ title, ...props }) => <Svg />

export default Test"
`;

exports[`reactNativeTemplate should wrap code into a component 7`] = `
"import React from 'react'
import Svg from 'react-native-svg';
const Test = ({ svgRef, ...props }) => <Svg />

export default Test"
`;
